### PR TITLE
Gives fridges a pretty UI with organizing features and image previews (alt title: FUCK IE7)

### PR DESCRIPTION
**Pull request by 9600bauds**
mpc-hc64_2016-12-10_13-28-03
dreamseeker_2016-12-10_13-26-20
https://a.uguu.se/wkpGGE7MZtWY_Desktop12.10.2016-13.15.37.02.webm

I spent way too much fucking time trying to make this pretty in FUCK IE7 FUCK IE7 FUCK IE7
:cl:
 * rscadd: Fridges now have a fancy new UI.

### CODE REVIEW

**Comment by PJB3005** [comment]
> No browser datums

Also, using IE8 is the minimum, not 7.

**Comment by 9600bauds** [comment]
Make a browser datum guide
My BYOND seemed to be stuck on IE7 for some reason
dreamseeker_2016-12-10_01-31-14

**Comment by PJB3005** [comment]
The user agent isn't correct.

**Comment by 9600bauds** [comment]
Well, fuck IE8 then
It can't even do absolute positioning with more than one argument
There is no fucking reason positioning something at the top right of a window should be this hackish

**Comment by PJB3005** [comment]
Ok better idea.

Nano

**Comment by 9600bauds** [comment]
no

**Comment by PJB3005** [comment]
Yes.

**Comment by 9600bauds** [comment]
this is not negotiable

**Comment by kilozombie** [comment]
Hold up, I recognize those pills!
Did you make the chemistry cheatsheet? I made the other one, the UPS.

**Comment by Kurfursten** [comment]
Self merge when?

**Comment by PJB3005** [comment]
When he Nanos it.

**Comment by 9600bauds** [comment]
Dear PJB:

No, I will not port something that is already finished to a shitty laggy UI that everybody hates, randomly does not work for some players (and nobody knows why), have been TOLD to never port something to again in the past, and from previous experience doing similar PRs can assert is a major PITA to work with.

I will especially not do it for seemingly no reason.

If you want me to implement this with browser datums I will do it if you post a readme of sorts.

**Comment by Kammerjunk** [comment]
On a content-related note, haha I can't wait to grudge people for their colours that are not *my* colours.
As far as I'm concerned, it *looks* great 👍 
Edit: And I even get to grudge them for their shelf choices!

**Comment by ComicIronic** [comment]
lel nano.

**Comment by ComicIronic** [comment]
Are the screenshots from when you had 5 shelves? The last shelf has a down arrow.

**Comment by 9600bauds** [comment]
Yes I lowered that to 4 because 3 is not enough and 5 is too much

**Comment by PJB3005** [comment]
```dm
var/datum/browser/fart = new(user, "\ref[src]", "fridge thing", width, height, src)
fart.set_content(html)
fart.open()
```

**Comment by 9600bauds** [comment]
What's the dealio with set_machine(src)? Do I need to put it somewhere/do I need to unset machine somewhere?

**Comment by PJB3005** [comment]
`set_machine` is shitcode and a complete mess I'd rather not get into to be honest.
It's an oldcode way of tracking what machine a mob is using (of course, oldcoders were too blind to realize multitasking is a thing).

If you were to use Nano you wouldn't have to fuck around with this shit but you're being a stubborn bastard so you'll have to write the shitcode related to managing updating yourself.
Why the hell do you think I advocate Nano? It's literally the best we've got and in the land of the blind the one-eyed is king.

Basically, you need to call `set_machine(src)` on the on users viewing the UI.
Then to update the UI you need to call `updateUsrDialog()`, this'll call `attack_hand()` again on every mob that's using the machine (but of course with the single machine limitation due to not using Nano) allowing you to update the UI for every client that's viewing it.

Don't worry about `unset_machine`, the browser datum handles it.

**Comment by 9600bauds** [comment]
Actually the whole debacle about updating shit was as simple as a set_machine(src) in attack_hand(), and src.updateUsrDialog() in Topic(), all of which was already there so woo

**Comment by PJB3005** [comment]
Did you really nuke the default browser datum CSS...?

**Comment by 9600bauds** [comment]
datum/browser/clean/ is not clean at all

**Comment by PJB3005** [comment]
No but God damnit man literally the reason I told you to use browser Datums was to make the UI not look like a turd and CONSISTENT with other interfaces *and you blew it*.

**Comment by 9600bauds** [comment]
Not a fan

**Comment by PJB3005** [comment]
Then fix your CSS so it fits with the regular stylesheet.

**Comment by 9600bauds** [comment]
oh goooodddddd was this all about the fucking dark CSS

**Comment by PJB3005** [comment]
Yes.

**Comment by 9600bauds** [comment]
Man I should be fucking studying and yet, here I am, dealing with you who can't write a sentence saying "i don't like how it looks" and instead try to boss me around and get me to port shit around just to BAIT ME into changing the color.

What the fuck PJB?

**Comment by 9600bauds** [comment]
firefox_2016-12-12_01-56-34

unless someone has something to say about the actual code this is done

**Comment by PJB3005** [comment]
This is not negotiable.

There are interfaces where not following the standard (standard being nano/default browser datum) makes sense. This is not such a case.

Would you think it fits if I suddenly made a new part of medbay and made it pitch black without context? No.

File: code/game/machinery/kitchen/smartfridge.dm [hunk 1]
```diff
@@ -13,11 +15,9 @@
 	flags = NOREACT
 	var/icon_on = "smartfridge"
 	var/icon_off = "smartfridge-off"
-	var/item_quants = list()
-	//var/ispowered = 1 //starts powered
-	//var/isbroken = 0
-	//OBSOLETE: That's what the BROKEN and NOPOWER stat bitflags are for
+	var/list/datum/fridge_pile/piles = list()
```
Comment (PJB3005) [inline code/game/machinery/kitchen/smartfridge.dm:18]: Where is the code that actually deletes these piles on deletion of the fridge itself?

File: code/game/machinery/kitchen/smartfridge.dm [hunk 2]
```diff
@@ -36,6 +36,30 @@
...
+	fridge.piles -= src.name
+	fridge = null
+
+/datum/fridge_pile/proc/addAmount(var/amt)
+	amount += amt
+
+/datum/fridge_pile/proc/removeAmount(var/amt)
+	amount -= amt
+	if(amount <= 0)
+		qdel(src)
```
Comment (PJB3005) [inline code/game/machinery/kitchen/smartfridge.dm]: Pool these.

File: code/game/machinery/kitchen/smartfridge.dm [hunk 3]
```diff
@@ -292,53 +317,117 @@
...
-	return
+		var/imagedesc
+		switch(display_miniicons)
+			if(MINIICONS_ON)
+				imagedesc = "On"
+			if(MINIICONS_UNCROPPED)
+				imagedesc = "Uncropped"
+			if(MINIICONS_OFF)
+				imagedesc = "Off"
+		dat += "<span style='position:absolute;right:2.5%;'><TT>Images: <a href='byond://?src=\ref[src];display_miniicons=1;'>[imagedesc]</A></TT></span>"
```
Comment (PJB3005) [inline code/game/machinery/kitchen/smartfridge.dm]: No inline CSS.

**Review by PJB3005 — DISMISSED** [review]
Couple more issues:

**Make this goddamn UI use the regular stylesheet.**

Do these piles work on name only? If I'm reading this right putting 10 poisonous grapes on shelf 1 then healthy ones on shelf 2 (same name), then taking from shelf 2 will get you the poisonous ones from shelf 1, which sort of defeats a ton of the use this has.

**Comment by 9600bauds** [comment]
>Would you think it fits if I suddenly made a new part of medbay and made it pitch black without context? No.

This is [literally what you're telling me to do.](http://i.imgur.com/2rZBjXe.png)

**Comment by 9600bauds** [comment]
>Do these piles work on name only? If I'm reading this right putting 10 poisonous grapes on shelf 1 then healthy ones on shelf 2 (same name), then taking from shelf 2 will get you the poisonous ones from shelf 1, which sort of defeats a ton of the use this has.

This is how it used to work forever. Out of scope.

**Comment by PJB3005** [comment]
Alright then, though I feel it's really counter intuitive to have them still be the "same" now that there's even different icons and shelves that would logically cause separation.

> This is literally what you're telling me to do.

Except the issue is, *your* UI stands out like a sore thumb compared to the rest of the UIs. Yes to the rest of *medbay* it might stand out but UIs shouldn't be compared against the environment they appear in. They're an entirely different part of the game.

**Comment by 9600bauds** [comment]
Addressed the code concerns, I will not be changing the styling.

File: code/game/machinery/kitchen/smartfridge.dm (hunk 1 above)
Comment (9600bauds) [reply code/game/machinery/kitchen/smartfridge.dm:18]: Added

**Comment by JustSumBody** [comment]
If only we had some central user interface system unique to VGstation
Maybe call it VGinterface?

**Comment by Intigracy** [comment]
WYCI with actual documentation

**Comment by Exxion** [comment]
OK I couldn't really give a rat's ass about the style, but the retrieval issue actually is relevant

**Comment by Intigracy** [comment]
I remember when the UI label was Nano You Why

**Comment by PJB3005** [comment]
@JustSumGuy I was working on that but gave up due to IE8.

**Comment by Exxion** [comment]
I think that was the joke

**Comment by 9600bauds** [comment]
What do you mean by retrieval issue @Exxion?

**Comment by Exxion** [comment]
>Do these piles work on name only? If I'm reading this right putting 10 poisonous grapes on shelf 1 then healthy ones on shelf 2 (same name), then taking from shelf 2 will get you the poisonous ones from shelf 1, which sort of defeats a ton of the use this has.

I get that this is how it has always worked, but the fact that it worked this way was never really relevant without the ability to split it into different shelves, because there could only be one entry per name anyway.

Unless there can still only be one entry per name, even across shelves, in which case yeah it is out of scope

**Comment by 9600bauds** [comment]
Yeah all the entries go by name. To separate grapes by shelf some method of adding something to a specific pile would have to be added, and there's nothing really intuitive besides something like having the fridge open with your fruit in hand and manually select the "active shelf" or something. Then the piles would have to merge anyways if they get moved into the same shelf.

For a while now we've had non-standard plants show up with an actually different names, like "modified grapes", "mutant grapes" and "grapes?". Yeah, all your modified grapes will be bunched together, but for the context of the fridge they will be separate from the standard at least.

**Comment by 9600bauds** [comment]
Just realized what PJB was trying to say - yeah you can't separate items with the exact same name into 2 different shelves

**Comment by Wizardcrying** [comment]
Labelling shelves?

**Comment by 9600bauds** [comment]
Maybe next time

**Comment by Probe1** [comment]
Resolve the two open comments @PJB3005 left about pooling please.

**Comment by 9600bauds** [comment]
I think I did

I don't see open comments

You sure that's not the "show outdated" thing?

**Comment by PJB3005** [comment]
Still disagree with not using standard styling.

Get with the times grandpa.

**Comment by 9600bauds** [comment]
firefox_2016-12-18_22-24-37

**Comment by Probe1** [comment]
image

delete your fUCKING bRANCHES OMG

**Comment by 9600bauds** [comment]
I'm still mad as fuck that I lost all my previous branches
https://github.com/9600bauds/vgstation13-old

**Comment by Probe1** [comment]
I'm horrified of you.

The new style isn't widely enough incorporated because we don't do it when we update shit.  It's not fun and it's tedious and frustrating.    Realistically, who next is going to come along and say "Oh I'll update it for you all".  Please consider changing to the new style so we can move forward with our UI instead of staying half in two styles?

**Comment by 9600bauds** [comment]
Black background is a pest

A pest I tell you

If anything I'm sooner to add an alternate white style and a client preference

**Comment by Probe1** [comment]
Could you do that after updating the style for this PR?

Please?

**Comment by 9600bauds** [comment]
you don't understand how much I want to be done with this PR

**Comment by Probe1** [comment]
I would not prostrate myself and say please if I thought there was any shadow of a hint of a ghost of a chance, that you would enjoy doing it in the least bit.

I know this is a shitty ask.  I know I can't really do anything to compensate you or convince you it's worth your personal time.  

But the server deserves unified UI and the server is worth me asking this of you.

**Comment by Probe1** [comment]
Also you can take this as a rainy day huge favor that I will be ethically and spiritually obliged to debt you with and you can cash it in on anything and I will do it.

**Comment by Shadowmech88** [comment]
*muffled "antag datums" in the distance*

**Comment by Probe1** [comment]
james-meme

**Comment by PJB3005** [comment]
> black background a pest

Better than eye burning white.

**Review by Probe1 — APPROVED** [review]
I would like it if we could agree on style but in this case we can't and the upgrade to functionality is more important.

**Comment by Probe1** [comment]
I pinged you all @Exxion @ComicIronic @clusterfack @Intigracy because we can't decide whether to merge it or hold out for bauds to change the style.  Cast tie breaking votes with review approval / needs changes please.

**Comment by Exxion** [comment]
It would be mathematically difficult for me to care less

**Review by Exxion — COMMENTED** [review]
*shrug

**Comment by Probe1** [comment]
Well clusterfacks dead, intigracy and exxion don't care, comics here then there then not seen anywhere.

@PJB3005 ve vill have to bite ze bullet and update the style tomorrow.
